Dynamic Dashboard: Reviews card states and unit tests

hafizrahman opened the pull request:
<!-- Remember about a good descriptive title. -->
Please do not merge until target is trunk

Closes: #12742
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

This PR completes the Reviews card task by adding the following:
1. Finalizing UI for loading state,
2. UI for error state,
3. UI for empty state (both for when there's no reviews at all in the site, and when there's no reviews after filtering). The wording here follows [Android's implementation](https://github.com/woocommerce/woocommerce-android/pull/11491/files#diff-b931a16277ac6133750e3d68c243da133842255a53e451802acd40aad3645e81R423-R424).
4. Dismiss and onError retry trackings
5. Unit tests

## Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->

Please check and compare with screenshots posted below:

1. Start with no internet, ensure "error state" is shown.
2. Retry, ensure this is tracked: `Tracked dynamic_dashboard_card_retry_tapped, ...` with `type: reviews` included.
3. Restart test, now with internet on a site with no reviews. ensure "Empty state case: absolutely no reviews" case shown.
4. Add one review on the site, but do not approve it. Go back to site, filter by "Approved". Ensure "Empty state case: no filtered reviews" case is shown.
5. Try reloading and ensure loading state is shown correctly.
6. HIde the card, ensure this is tracked: `Tracked dynamic_dashboard_hide_card_tapped, ...` with `type: reviews` included.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->

### Error case:
 <img src="https://github.com/woocommerce/woocommerce-ios/assets/266376/a7d5e960-eeef-44df-b0b1-eec49d5153b2" width="42%"> 

| Empty state case: absolutely no reviews | Empty state case:  no filtered reviews |
|-|-|
|![Simulator Screenshot - iPhone 15 iOS 17 4 - 2024-05-22 at 09 42 07](https://github.com/woocommerce/woocommerce-ios/assets/266376/66e14eea-d587-4b2c-9e14-e0d046115054) | ![Simulator Screenshot - iPhone 15 iOS 17 4 - 2024-05-22 at 09 39 49](https://github.com/woocommerce/woocommerce-ios/assets/266376/98613b44-d4a8-48a9-8a8f-a2603368c2fd) |


| Loading state case: no reviews ever loaded | Loading state case: some reviews were shown before |
|-|-|
| ![Simulator Screenshot - iPhone 15 iOS 17 4 - 2024-05-20 at 16 54 32](https://github.com/woocommerce/woocommerce-ios/assets/266376/de7ef8c3-8984-488d-8fe8-ac52e56e8bf4) | ![Simulator Screenshot - iPhone 15 iOS 17 4 - 2024-05-20 at 16 54 21](https://github.com/woocommerce/woocommerce-ios/assets/266376/3a37992b-0b51-4f58-bb68-8700e744cc67)  |

hafizrahman committed:
Add empty state.

hafizrahman committed:
Add error handling

hafizrahman committed:
Add dismiss tracking

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Merge branch 'feat/12742-review-card-functionality-ii'

hafizrahman committed:
Add unit tests.

hafizrahman committed:
Hide filter divider when loading to reduce extra line showing.

hafizrahman committed:
Update empty state wordings matching Android.

dangermattic commented:
<!--
  1 Error: This PR is tagged with <code>s...
  0 Warnings
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">This PR is tagged with <code>status: do not merge</code> label(s).</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCard.swift:
```diff
@@ -209,6 +222,13 @@ private extension ReviewsDashboardCard {
+
+    func emptyView(isFiltered: Bool) -> some View {
+        VStack(spacing: 0) {
+            ReviewsDashboardEmptyView(isFiltered: isFiltered)
```
Nit: this view is simple so can be part of this file, not necessary to be separate unless it's to be reused.

hafizrahman replied:
good point, updated in https://github.com/woocommerce/woocommerce-ios/pull/12806/commits/811f143498d9f08ec28e692a662dbdb5edc614e4

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardEmptyView.swift:
```diff
@@ -0,0 +1,39 @@
+
+        static let noFilteredReviewsText = NSLocalizedString(
+            "mostActiveCouponsEmptyView.noFilteredReviewsText",
+            value: "No reviews match the selected filter, please try changing the filter.",
```
Nit: if this view is kept in `ReviewDashboardCard`, you can use the current filter to display in the message. Something like: "No reviews matching Hold status". This would be more explicit and concise.

hafizrahman replied:
Yup, that's more useful. Added in https://github.com/woocommerce/woocommerce-ios/pull/12806/commits/811f143498d9f08ec28e692a662dbdb5edc614e4

itsmeichigo commented on WooCommerce/WooCommerceTests/ViewRelated/Dashboard/Reviews/ReviewsDashboardCardViewModelTests.swift:
```diff
@@ -0,0 +1,307 @@
+        await viewModel.reloadData()
+
+        // Then
+        XCTAssertEqual(viewModel.data.count, self.sampleReviews.count)
```
We should compare the data rather than just the count, to make sure that the details are correct.

hafizrahman replied:
Thanks, updated in https://github.com/woocommerce/woocommerce-ios/pull/12806/commits/8739e01865f91f53a138a6f3aa2bd12f9de5597f

hafizrahman committed:
Merge branch 'trunk'

Conflicts:
	WooCommerce/Classes/ViewRelated/Dashboard/Reviews/ReviewsDashboardCard.swift

hafizrahman committed:
Fix post-merge issue.

hafizrahman committed:
Fix some broken unit tests.

hafizrahman committed:
Refactor to update data using fetchedReviews as it becomes available.

hafizrahman committed:
Merge branch 'trunk'

Conflicts:
	WooCommerce/WooCommerce.xcodeproj/project.pbxproj

hafizrahman committed:
Refactor empty view to simplify and add filter name in the empty view message.

hafizrahman committed:
Add unit test for switchingStatus

itsmeichigo reviewed (commented):
Hey @hafizrahman 👋  As discussed in the previous PR, I think we should not always show the review list when there is existing data. From my testing, when refreshing the card, we show the list without the shimmering animation, making it confusing to understand if data is being reloaded.

hafizrahman commented:
@itsmeichigo I think something might have gone wrong / unaccounted for after merging from the latest branch, I was just responding to the previous feedbacks without adding any logic change. Let me check, thanks!

hafizrahman committed:
Revamp loading state to show shimmering placeholder data always.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/206">18.9</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hafizrahman committed:
Simplify loading to just use syncingData

hafizrahman commented:
@itsmeichigo I decided to revamp the loading state to match similar behavior in InboxDashboardCard and have placeholder data displayed always when loading. I also simplified the syncing to only show items when there's full text with product names and notifications loaded, just makes things easier to follow that way.

Added in [b758898](https://github.com/woocommerce/woocommerce-ios/pull/12806/commits/b75889894c7bcb7391d9b0be8c8e485a4741e316) and [8d87f7f](https://github.com/woocommerce/woocommerce-ios/pull/12806/commits/8d87f7fdce4d7752557ff4fcdc9854cc113bf181)

Video of latest loading state:


https://github.com/woocommerce/woocommerce-ios/assets/266376/7128474e-3004-4fb8-aed8-37e0611249d9

hafizrahman committed:
Update unit tests

itsmeichigo reviewed (approved):
Thanks Hafiz, LGTM :shipit:
